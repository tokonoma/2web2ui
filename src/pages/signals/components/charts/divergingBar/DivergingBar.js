import React, { Component } from 'react';
import _ from 'lodash';
import PropTypes from 'prop-types';
import { tokens } from '@sparkpost/design-tokens-hibana';
import {
  ResponsiveContainer,
  BarChart,
  Bar,
  Tooltip,
  XAxis,
  YAxis,
  CartesianGrid,
  Rectangle,
  Text,
} from 'recharts';
import useHibanaOverride from 'src/hooks/useHibanaOverride';
import TooltipWrapper from 'src/components/charts/Tooltip';
import OGStyles from './DivergingBar.module.scss';
import hibanaStyles from './DivergingBarHibana.module.scss';

export class DivergingBarClassComponent extends Component {
  getData = () => {
    const { data, xKey, positiveFill, negativeFill } = this.props;
    return (
      data && data.map(item => ({ ...item, fill: item[xKey] > 0 ? positiveFill : negativeFill }))
    );
  };

  getHeight = () => {
    const { data, barHeight } = this.props;
    return data && data.length * barHeight;
  };

  renderBar = ({ key, background, payload, ...props }) => {
    const { selected, yKey } = this.props;
    const width = background.x + background.width;
    return (
      <g>
        <rect
          {...props}
          key={`${key}-bg`}
          x={0}
          x1={0}
          x2={width}
          width={width}
          fill={tokens.color_blue_400}
          opacity={selected && selected === payload[yKey] ? 0.3 : 0}
        />
        <Rectangle key={key} {...props} />
      </g>
    );
  };

  renderYTick = ({ payload, ...props }) => {
    const { data, selected, yKey, yLabel } = this.props;
    const match = _.find(data, [yKey, selected]) || {};
    const label = yLabel ? yLabel(payload) : payload.value;

    if (payload.value === match[yKey]) {
      return (
        <Text {...props} fill={tokens.colors_gray_800}>
          {label}
        </Text>
      );
    }

    return <Text {...props}>{label}</Text>;
  };

  render() {
    const { tooltipContent, onClick, width, xDomain, xKey, yKey, styles } = this.props;

    return (
      <ResponsiveContainer height={this.getHeight()} width={width} className={styles.DivergingBar}>
        <BarChart data={this.getData()} layout="vertical" barCategoryGap={2}>
          <CartesianGrid
            horizontal={false}
            shapeRendering="crispEdges"
            stroke={tokens.color_gray_400}
          />
          <Bar
            cursor="pointer"
            dataKey={xKey}
            onClick={onClick}
            isAnimationActive={false}
            shape={this.renderBar}
            minPointSize={1}
          />
          <YAxis
            type="category"
            tickLine={false}
            axisLine={false}
            interval={0}
            dataKey={yKey}
            padding={{ bottom: 5 }}
            tick={this.renderYTick}
            width={160}
          />
          <XAxis
            hide
            type="number"
            tickLine={false}
            domain={xDomain}
            dataKey={xKey}
            shapeRendering="crispEdges"
            ticks={[0]}
          />
          <Tooltip
            cursor={false}
            isAnimationActive={false}
            content={<TooltipWrapper children={tooltipContent} />}
            position={{ x: 0, y: 0 }}
          />
        </BarChart>
      </ResponsiveContainer>
    );
  }
}

// TODO: Remove when OG theme is removed
function DivergingBar(props) {
  const styles = useHibanaOverride(OGStyles, hibanaStyles);

  return <DivergingBarClassComponent {...props} styles={styles} />;
}

DivergingBar.propTypes = {
  negativeFill: PropTypes.string,
  onClick: PropTypes.func,
  positiveFill: PropTypes.string,
  tooltipContent: PropTypes.func,
  xDomain: PropTypes.array,
  xKey: PropTypes.string,
  yKey: PropTypes.string,
  yLabel: PropTypes.func,
  width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
};

DivergingBar.defaultProps = {
  barHeight: 35,
  xDomain: ['dataMin', 'dataMax'],
  xKey: 'value',
  yKey: 'label',
  width: '99%',
  negativeFill: tokens.color_red_700,
  positiveFill: tokens.color_green_700,
};

export default DivergingBar;
