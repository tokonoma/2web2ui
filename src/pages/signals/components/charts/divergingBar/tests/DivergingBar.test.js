import React from 'react';
import { shallow } from 'enzyme';
import { DivergingBarClassComponent as DivergingBar } from '../DivergingBar';
import styles from '../DivergingBar.module.scss';

describe('DivergingBar Component', () => {
  let wrapper;
  let props;
  const data = [
    { label: 'Label foo', value: '0.5' },
    { label: 'Label bar', value: '-0.5' },
  ];
  const getPayload = i => ({
    key: data[i].key,
    background: { x: 0, width: 100 },
    payload: { ...data[i], value: data[i].label },
  });

  beforeEach(() => {
    props = {
      barHeight: 10,
      data,
      xKey: 'value',
      yKey: 'label',
      width: '50%',
      onClick: jest.fn(),
      tooltipContent: jest.fn(),
      styles,
    };
    wrapper = shallow(<DivergingBar {...props} />);
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle click on Bar', () => {
    wrapper.find('Bar').simulate('click');
    expect(props.onClick).toHaveBeenCalled();
  });

  it('should render custom unselected bar', () => {
    const Bar = wrapper
      .find('Bar')
      .at(0)
      .props()
      .shape(getPayload(0));
    expect(Bar).toMatchSnapshot();
  });

  it('should render selected bar', () => {
    wrapper.setProps({ selected: data[0].label });
    const Bar = wrapper
      .find('Bar')
      .props()
      .shape(getPayload(0));
    expect(Bar).toMatchSnapshot();
  });

  it('should render custom unselected y tick', () => {
    const Text = wrapper
      .find('YAxis')
      .props()
      .tick(getPayload(0));
    expect(Text).toMatchSnapshot();
  });
});
