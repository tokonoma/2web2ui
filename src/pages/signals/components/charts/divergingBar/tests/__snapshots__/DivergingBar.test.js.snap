// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DivergingBar Component renders correctly 1`] = `
<ResponsiveContainer
  className="DivergingBar"
  debounce={0}
  height={20}
  width="50%"
>
  <BarChart
    barCategoryGap={2}
    barGap={4}
    data={
      Array [
        Object {
          "fill": undefined,
          "label": "Label foo",
          "value": "0.5",
        },
        Object {
          "fill": undefined,
          "label": "Label bar",
          "value": "-0.5",
        },
      ]
    }
    layout="vertical"
    margin={
      Object {
        "bottom": 5,
        "left": 5,
        "right": 5,
        "top": 5,
      }
    }
    reverseStackOrder={false}
    stackOffset="none"
  >
    <CartesianGrid
      fill="none"
      horizontal={false}
      horizontalFill={Array []}
      horizontalPoints={Array []}
      shapeRendering="crispEdges"
      stroke="#c5ced6"
      vertical={true}
      verticalFill={Array []}
      verticalPoints={Array []}
    />
    <Bar
      animationBegin={0}
      animationDuration={400}
      animationEasing="ease"
      cursor="pointer"
      data={Array []}
      dataKey="value"
      hide={false}
      isAnimationActive={false}
      layout="vertical"
      legendType="rect"
      minPointSize={1}
      onAnimationEnd={[Function]}
      onAnimationStart={[Function]}
      onClick={[MockFunction]}
      shape={[Function]}
      xAxisId={0}
      yAxisId={0}
    />
    <YAxis
      allowDataOverflow={false}
      allowDecimals={true}
      allowDuplicatedCategory={true}
      axisLine={false}
      dataKey="label"
      domain={
        Array [
          0,
          "auto",
        ]
      }
      height={0}
      hide={false}
      interval={0}
      mirror={false}
      orientation="left"
      padding={
        Object {
          "bottom": 5,
        }
      }
      reversed={false}
      scale="auto"
      tick={[Function]}
      tickCount={5}
      tickLine={false}
      type="category"
      width={160}
      yAxisId={0}
    />
    <XAxis
      allowDataOverflow={false}
      allowDecimals={true}
      allowDuplicatedCategory={true}
      dataKey="value"
      domain={
        Array [
          0,
          "auto",
        ]
      }
      height={30}
      hide={true}
      mirror={false}
      orientation="bottom"
      padding={
        Object {
          "left": 0,
          "right": 0,
        }
      }
      reversed={false}
      scale="auto"
      shapeRendering="crispEdges"
      tickCount={5}
      tickLine={false}
      ticks={
        Array [
          0,
        ]
      }
      type="number"
      width={0}
      xAxisId={0}
    />
    <Tooltip
      active={false}
      animationDuration={400}
      animationEasing="ease"
      content={
        <Tooltip
          offset={25}
          width="200px"
        >
          [MockFunction]
        </Tooltip>
      }
      contentStyle={Object {}}
      coordinate={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      cursor={false}
      cursorStyle={Object {}}
      filterNull={true}
      isAnimationActive={false}
      itemSorter={[Function]}
      itemStyle={Object {}}
      labelStyle={Object {}}
      offset={10}
      position={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      separator=" : "
      useTranslate3d={false}
      viewBox={
        Object {
          "x1": 0,
          "x2": 0,
          "y1": 0,
          "y2": 0,
        }
      }
      wrapperStyle={Object {}}
    />
  </BarChart>
</ResponsiveContainer>
`;

exports[`DivergingBar Component should render custom unselected bar 1`] = `
<g>
  <rect
    fill="#abd2ff"
    opacity={0}
    width={100}
    x={0}
    x1={0}
    x2={100}
  />
  <Rectangle
    animationBegin={0}
    animationDuration={1500}
    animationEasing="ease"
    height={0}
    isAnimationActive={false}
    isUpdateAnimationActive={false}
    radius={0}
    width={0}
    x={0}
    y={0}
  />
</g>
`;

exports[`DivergingBar Component should render custom unselected y tick 1`] = `
<Text
  background={
    Object {
      "width": 100,
      "x": 0,
    }
  }
  capHeight="0.71em"
  lineHeight="1em"
  scaleToFit={false}
  textAnchor="start"
  verticalAnchor="end"
  x={0}
  y={0}
>
  Label foo
</Text>
`;

exports[`DivergingBar Component should render selected bar 1`] = `
<g>
  <rect
    fill="#abd2ff"
    opacity={0.3}
    width={100}
    x={0}
    x1={0}
    x2={100}
  />
  <Rectangle
    animationBegin={0}
    animationDuration={1500}
    animationEasing="ease"
    height={0}
    isAnimationActive={false}
    isUpdateAnimationActive={false}
    radius={0}
    width={0}
    x={0}
    y={0}
  />
</g>
`;
